src/include/primer/p0_trie.h

```diff
@@ -0,0 +1,312 @@
+   * @param key_char Key char of child node.
+   * @return True if this trie node has a child with given key, false otherwise.
+   */
+  bool HasChild(char key_char) const { return false; }
+
+  /**
+   * TODO(P0): Add implementation
+   *
+   * @brief Whether this trie node has any children at all. This is useful
+   * when implementing 'Remove' functionality.
+   *
+   * @return True if this trie node has any child node, false if it has no child node.
+   */
+  bool HasChildren() const { return false; }
+
+  /**
+   * TODO(P0): Add implementation
+   *
+   * @brief Whether this trie node is the ending character of a key string.
+   *
+   * @return True if is_end_ flag is true, false if is_end_ is false.
+   */
+  bool IsEndNode() const { return false; }
+
+  /**
+   * TODO(P0): Add implementation
+   *
+   * @brief Return key char of this trie node.
+   *
+   * @return key_char_ of this trie node.
+   */
+  char GetKeyChar() const { return 'a'; }
+
+  /**
+   * TODO(P0): Add implementation
+   *
+   * @brief Insert a child node for this trie node into children_ map, given the key char and
+   * unique_ptr of the child node. If specified key_char already exists in children_,
+   * return nullptr. If parameter `child`'s key char is different than parameter
+   * `key_char`, return nullptr.
+   *
+   * Note that parameter `child` is rvalue and should be moved when it is
+   * inserted into children_map.
+   *
+   * The return value is a pointer to unique_ptr because pointer to unique_ptr can access the
+   * underlying data without taking ownership of the unique_ptr. Further, we can set the return
+   * value to nullptr when error occurs.
+   *
+   * @param key Key of child node
+   * @param child Unique pointer created for the child node. This should be added to children_ map.
+   * @return Pointer to unique_ptr of the inserted child node. If insertion fails, return nullptr.
+   */
+  std::unique_ptr<TrieNode> *InsertChildNode(char key_char, std::unique_ptr<TrieNode> &&child) { return nullptr; }
+
+  /**
+   * TODO(P0): Add implementation
+   *
+   * @brief Get the child node given its key char. If child node for given key char does
+   * not exist, return nullptr.
+   *
+   * @param key Key of child node
+   * @return Pointer to unique_ptr of the child node, nullptr if child
+   *         node does not exist.
+   */
+  std::unique_ptr<TrieNode> *GetChildNode(char key_char) { return nullptr; }
+
+  /**
+   * TODO(P0): Add implementation
+   *
+   * @brief Remove child node from children_ map.
+   * If key_char does not exist in children_, return immediately.
+   *
+   * @param key_char Key char of child node to be removed
+   */
+  void RemoveChildNode(char key_char) {}
+
+  /**
+   * TODO(P0): Add implementation
+   *
+   * @brief Set the is_end_ flag to true or false.
+   *
+   * @param is_end Whether this trie node is ending char of a key string
+   */
+  void SetEndNode(bool is_end) {}
+
+ protected:
+  /** Key character of this trie node */
```

Comment:
May also remove all member variables if we don't want to put too much constraint on students.